ctapipe/io/serializer.py

```diff
@@ -0,0 +1,356 @@
+"""
+Serialize ctapipe containers to file
+"""
+
+from astropy.table import Table, Column
+from ctapipe.core import Container
+from ctapipe.core import Component
+from abc import ABC, abstractmethod
+from astropy import log
+from pickle import load
+from pickle import dump
+from pickle import PickleError
+from traitlets import Unicode
+import numpy as np
+import gzip
+
+__all__ = ['FitsSerializer', 'PickleSerializer']
+log.setLevel('INFO')
+
+
+# PICKLE GZIP Implementation
+
+class Serializer(ABC, Component):
+    """
+    Base class for context manager to save Containers to file
+    """
+    def __init__(self, file,  overwrite=False):
+        """
+        Parameters
+        ----------
+        file:  Unicode
+            input or output full path file name
+        overwrite: Bool
+            overwrite outfile file if it already exists
+        """
+        self.file = file
+        self.overwrite = overwrite
+        super().__init__()
+
+    def __enter__(self):
+        log.debug("Serializing on {0}".format(self.file))
+        return self
+
+    def __exit__(self, *args):
+        pass
+
+    @abstractmethod
+    def add_container(self, container):
+        """
+        Add a container to serializer
+        """
+        pass
+
+    @abstractmethod
+    def write(self):
+        """
+        Write all containers to outfile
+        """
+        pass
+
+    @abstractmethod
+    def __iter__(self):
```

Comment:
Ah, sorry. Now I get It. You mixed serialization and deserialisation. I think we should keep that seperate. Having writers and readers. We sources in `io`, correct? 

So maybe a PickleSource would be a better idea.

Especially: what happens if I mix iteration with `add_container`? I think this should not be possible and an easy solution would be to have seperate classes for readers and writers.